File: src/games/stendhal/client/ImageSprite.java
/*
 * @(#) src/games/stendhal/client/ImageSprite.java
 *
 * $Id$
 */

package games.stendhal.client;

//
//
import java.awt.Color;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.GraphicsConfiguration;
import java.awt.GraphicsEnvironment;
import java.awt.Image;
import java.awt.Transparency;
import java.awt.image.BufferedImage;

/**
 * A sprite to be displayed on the screen. Note that a sprite contains no state
 * information, i.e. its just the image and not the location. This allows us to
 * use a single sprite in lots of different places without having to store
 * multiple copies of the image.
 * 
 * @author Kevin Glass
 */
public class ImageSprite implements Sprite {
	/** a brighter version of the sprite */
	protected ImageSprite brighterSprite;

	/** a darker version of the sprite */
	protected ImageSprite darkerSprite;

	/** The image to be drawn for this sprite */
	protected Image image;

	/**
	 * The identifier reference.
	 */
	protected Object	reference;


	/**
	 * Create a new sprite based on an image
	 * 
	 * @param image
	 *            The image that is this sprite
	 */
	public ImageSprite(Image image) {
		this(image, null);
	}


	/**
	 * Create a new sprite based on an image
	 * 
	 * @param	image		The image that is this sprite.
	 * @param	reference	The sprite reference, or null.
	 */
	public ImageSprite(Image image, String reference) {
		this.image = image;
		this.reference = reference;
	}


	public ImageSprite(Sprite sprite) {
		this(sprite, null);
	}


	/**
	 * Create a copy of another sprite.
	 * 
	 * @param	sprite		The source sprite.
	 * @param	reference	The sprite reference, or null.
	 */
	public ImageSprite(Sprite sprite, String reference) {
		this.reference = reference;

		image = getGC().createCompatibleImage(sprite.getWidth(), sprite.getHeight(), Transparency.BITMASK);

		sprite.draw(image.getGraphics(), 0, 0);
	}


	//
	// ImageSprite
	//

	/**
	 * Create a sprite with the image flipped horizontally.
	 *
	 * @return	A horizontally flipped sprite.
	 */
	public static ImageSprite flipped(Sprite sprite) {
		Image image = getGC().createCompatibleImage(sprite.getWidth(), sprite.getHeight(), Transparency.BITMASK);

		int width = sprite.getWidth();

		sprite.draw(image.getGraphics(), width, 0, width, 0, -width, sprite.getHeight());

		return new ImageSprite(image);
	}


	protected static GraphicsConfiguration getGC() {
		return GraphicsEnvironment.getLocalGraphicsEnvironment().getDefaultScreenDevice().getDefaultConfiguration();
	}


	/**
	 * Get the graphics context of the underlying image.
	 *
	 * @return	The graphics context.
	 */
	public Graphics getGraphics() {
		return image.getGraphics();
	}


	/**
	 * Flip the image horizontally.
	 * @return an horizontal flipped sprite.
	 */
	public ImageSprite flip() {
		// TODO: Eventually, just call: flipped(this)

		Image empty = getGC().createCompatibleImage(getWidth(), getHeight(), Transparency.BITMASK);
		ImageSprite spr=new ImageSprite(empty);

		spr.getGraphics().drawImage(image, getWidth(), 0, 0, getHeight(), 0, 0, getWidth(), getHeight(), null);

		return spr;
	}


	/** overlays the image with the given color and returns a new image. */
	private Image getModifiedImage(Color color, float alpha) {
		BufferedImage i = getGC().createCompatibleImage(getWidth(), getHeight(), Transparency.TRANSLUCENT);
		draw(i.getGraphics(), 0, 0);
		Graphics2D g = i.createGraphics();
		g.setColor(new Color(color.getRed(), color.getGreen(), color.getBlue(), (int) (alpha * 255)));
		g.fillRect(0, 0, i.getWidth(), i.getHeight());

		return i;
	}

	/** returns a brighter version of the sprite */
	public synchronized ImageSprite brighter() {
		if (brighterSprite == null) {
			brighterSprite = new ImageSprite(getModifiedImage(Color.WHITE, 0.3f));
			brighterSprite.darkerSprite = this;
		}
		return brighterSprite;
	}

	/** returns a darker version of the sprite */
	public synchronized ImageSprite darker() {
		if (darkerSprite == null) {
			darkerSprite = new ImageSprite(getModifiedImage(Color.BLACK, 0.3f));
			darkerSprite.brighterSprite = this;
		}
		return darkerSprite;
	}


	//
	// Sprite
	//

	/**
	 * Copy the sprite.
	 *
	 * @return	A new copy of the sprite.
	 */
	public Sprite copy() {
		return new ImageSprite(this);
	}

	/**
	 * Draw the sprite onto the graphics context provided
	 * 
	 * @param g
	 *            The graphics context on which to draw the sprite
	 * @param x
	 *            The x location at which to draw the sprite
	 * @param y
	 *            The y location at which to draw the sprite
	 */
	public void draw(Graphics g, int x, int y) {
		g.drawImage(image, x, y, null);
	}

	/**
	 * Draws the image
	 * 
	 * @param g
	 *            the graphics context where to draw to
	 * @param destx
	 *            destination x
	 * @param desty
	 *            destination y
	 * @param x
	 *            the source x
	 * @param y
	 *            the source y
	 * @param w
	 *            the width
	 * @param h
	 *            the height
	 */
	// Bugfix: to use image.getWidth()/getHeight() is not correct for images
	// coming from the Tilestore, as those are used to draw more than 1 Sprite
	// from the same image. The relevant image size is that of the image
	// that we're painting in, but the Graphics context doesn't say anything
	// about the size of the image it belongs to so I had to add parameters
	// for width and height. This bug was responsible for the drawing problems
	// on Mac OS X.
	// What I don't understand now though is why it worked well on Windows ;)
	// This bugfix also affects Sprite.draw in TileStore.java and
	// SpriteStore.java
	// intensifly @ gmx.com, April 20th, 2006
	// public void draw(Graphics g, int destx, int desty, int x,int y) {
	public void draw(Graphics g, int destx, int desty, int x, int y, int w, int h) {
		g.drawImage(image, destx, desty, destx + w, desty + h, x, y, x + w, y + h, null);
	}

	/**
	 * Get the height of the drawn sprite
	 * 
	 * @return The height in pixels of this sprite
	 */
	public int getHeight() {
		return image.getHeight(null);
	}

	/**
	 * Get the sprite reference. This identifier is an externally
	 * opaque object that implements equals() and hashCode() to
	 * uniquely/repeatably reference a keyed sprite.
	 *
	 * @return	The reference identifier, or <code>null</code> if
	 *		not referencable.
	 */
	public Object getReference() {
		return reference;
	}

	/**
	 * Get the width of the drawn sprite
	 * 
	 * @return The width in pixels of this sprite
	 */
	public int getWidth() {
		return image.getWidth(null);
	}


	//
	// Object
	//

	@Override
	public boolean equals(Object obj) {
		if(obj instanceof ImageSprite) {
			ImageSprite img=(ImageSprite)obj;
			return image.equals(img.image);
		}

		return false;
	}

	@Override
	public int hashCode() {
		return image.hashCode();
	}
}

File: srcjs/stendhal/sprite/ImageSprite.ts
/***************************************************************************
 *                   (C) Copyright 2003-2023 - Stendhal                    *
 ***************************************************************************
 ***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU Affero General Public License as        *
 *   published by the Free Software Foundation; either version 3 of the    *
 *   License, or (at your option) any later version.                       *
 *                                                                         *
 ***************************************************************************/

import { EmptySprite } from "./EmptySprite";
import { Sprite } from "./Sprite";

/**
 * A sprite to be displayed on the screen. Note that a sprite contains no state
 * information, i.e. its just the image and not the location. This allows us to
 * use a single sprite in lots of different places without having to store
 * multiple copies of the image.
 *
 * @author Kevin Glass
 */
export class ImageSprite implements Sprite {

    /** The image to be drawn for this sprite. */
    private image: CanvasImageSource;

    /**
     * The identifier reference.
     */
    private reference?: object;

    /**
     * Create a new sprite based on an image.
     *
     * @param image
     *            The image that is this sprite.
     * @param reference
     *            The sprite reference, or null.
     */
    public constructor (image: CanvasImageSource | Sprite, reference?: object) {
        if (image instanceof Sprite) {
            this.image = ImageSprite.createCompatibleImage(image);
        } else {
            this.image = image;
        }
        this.reference = reference;
    }

    private static createCompatibleImage(sprite: Sprite): HTMLCanvasElement {
        const canvas = document.createElement('canvas');
        const g = canvas.getContext('2d');
        canvas.width = sprite.getWidth();
        canvas.height = sprite.getHeight();
        sprite.draw(g!, 0, 0);
        return canvas;
    }


    /**
     * Create a sub-region of this sprite. <strong>NOTE: This does not use
     * caching.</strong>
     *
     * @param x
     *            The starting X coordinate.
     * @param y
     *            The starting Y coordinate.
     * @param width
     *            The region width.
     * @param height
     *            The region height.
     * @param ref
     *            The sprite reference.
     *
     * @return A new sprite.
     */
    public createRegion(x: number, y: number, width: number, height: number, ref?: object): Sprite {
        const iwidth = this.getWidth();
        const iheight = this.getHeight();

        if ((x >= iwidth) || (y >= iheight)) {
            // Outside of image (nothing to draw)
            return new EmptySprite(width, height, ref);
        }

        // Exclude regions outside the original image
        width = Math.min(width, iwidth);
        height = Math.min(height, iheight);

        // Full copy method (the memory hog)
        const canvas = document.createElement('canvas');
        const g = canvas.getContext('2d');
        canvas.width = width;
        canvas.height = height;

        this.draw(g!, 0, 0, x, y, width, height);

        return new ImageSprite(canvas, ref);
    }

    /**
     * Draw the sprite onto the graphics context provided.
     *
     * @param g
     *            The graphics context on which to draw the sprite
     * @param x
     *            The x location at which to draw the sprite
     * @param y
     *            The y location at which to draw the sprite
     */
    public draw(g: CanvasRenderingContext2D, destx: number, desty: number): void;

    /**
     * Draws the image.
     *
     * @param g
     *            the graphics context where to draw to
     * @param destx
     *            destination x
     * @param desty
     *            destination y
     * @param x
     *            the source x
     * @param y
     *            the source y
     * @param w
     *            the width
     * @param h
     *            the height
     */
    public draw(g: CanvasRenderingContext2D, destx: number, desty: number, x: number, y: number, w: number, h: number): void;
    public draw(g: CanvasRenderingContext2D, destx: number, desty: number, x?: number, y?: number, w?: number, h?: number): void {
            if (arguments.length === 7) {
            g.drawImage(this.image, destx, desty, destx + w!, desty + h!, x!, y!, x! + w!, y! + h!);
        } else {
            g.drawImage(this.image, destx, desty);
        }
    }

    /**
     * Get the height of the drawn sprite.
     *
     * @return The height in pixels of this sprite
     */
    public getHeight(): number {
        return this.image.height as number;
    }

    /**
     * Get the sprite reference. This identifier is an externally opaque object
     * that implements equals() and hashCode() to uniquely/repeatably reference
     * a keyed sprite.
     *
     * @return The reference identifier, or <code>null</code> if not
     *         referencable.
     */
    public getReference(): object | undefined {
        return this.reference;
    }

    /**
     * Get the width of the drawn sprite.
     *
     * @return The width in pixels of this sprite
     */
    public getWidth(): number {
        return this.image.width as number;
    }

    public isConstant(): boolean {
        return true;
    }

    /**
     * Retrieves a single frame from the image.
     *
     * @param xIndex
     * 		Horizontal index.
     * @param yIndex
     * 		Vertical index.
     * @return
     * 		Cropped Sprite.
     */
    public getFrame(xIndex: number, yIndex: number): Sprite {
        const w = this.getWidth() / 3;
        const h = this.getHeight() / 4;
        const x = w * xIndex;
        const y = h * yIndex;

        return this.createRegion(x, y, w, h, undefined);
    }
}
